Sources/SmithyCBOR/Reader/Reader.swift

```diff
@@ -0,0 +1,237 @@
+                children.append(child)
+            }
+        default:
+            break
+        }
+        return children
+    }
+
+    public subscript(nodeInfo: NodeInfo) -> Reader {
+        if let match = children.first(where: { $0.nodeInfo == nodeInfo }) {
+            return match
+        } else {
+            return Reader(nodeInfo: nodeInfo, cborValue: nil, parent: self)
+        }
+    }
+
+    public func readIfPresent() throws -> String? {
+        switch cborValue {
+        case .text(let string):
+            return string
+        case .indef_text_start:
+            // Handle concatenation of indefinite-length text
+            var combinedText = ""
+            for child in children {
+                if let chunk = try child.readIfPresent() as String? {
+                    combinedText += chunk
+                }
+            }
+            return combinedText
+        case .bytes(let data):
+            return String(data: data, encoding: .utf8)
+        default:
+            return nil
+        }
+    }
+
+    public func readIfPresent() throws -> Int8? {
+        switch cborValue {
+        case .int(let intValue): return Int8(intValue)
+        case .uint(let uintValue): return Int8(uintValue)
+        default: return nil
+        }
+    }
+
+    public func readIfPresent() throws -> Int16? {
+        switch cborValue {
+        case .int(let intValue): return Int16(intValue)
+        case .uint(let uintValue): return Int16(uintValue)
+        default: return nil
+        }
+    }
+
+    public func readIfPresent() throws -> Int? {
+        switch cborValue {
+        case .int(let intValue): return Int(intValue)
+        case .uint(let uintValue): return Int(uintValue)
+        default: return nil
+        }
+    }
+
+    public func readIfPresent() throws -> Float? {
+        switch cborValue {
+        case .double(let doubleValue): return Float(doubleValue)
+        case .int(let intValue): return Float(intValue)
+        case .uint(let uintValue): return Float(uintValue)
+        default: return nil
+        }
+    }
+
+    public func readIfPresent() throws -> Double? {
+        switch cborValue {
+        case .double(let doubleValue): return doubleValue
+        case .int(let intValue): return Double(intValue)
+        case .uint(let uintValue): return Double(uintValue)
+        default: return nil
+        }
+    }
+
+    public func readIfPresent() throws -> Bool? {
+        switch cborValue {
+        case .bool(let boolValue): return boolValue
+        default: return nil
+        }
+    }
+
+    public func readIfPresent() throws -> Data? {
+        switch cborValue {
+        case .bytes(let data): return data
+        case .text(let string): return Data(base64Encoded: string)
+        default: return nil
+        }
+    }
+
+    public func readIfPresent() throws -> Document? {
+        guard let cborValue else { return nil }
+        return Document(cborValue as! SmithyDocument)
+    }
```

Comment:
Q: RPCv2 CBOR SEP says Smithy's document type isn't yet supported in this protocol; should this just return nil?

Reply:
oh yeah will change, in write its just empty